src/lgdo/types/histogram.py

```diff
@@ -315,6 +324,98 @@ def binning(self) -> tuple[Histogram.Axis, ...]:
         assert all(isinstance(v, Histogram.Axis) for k, v in bins)
         return tuple(v for _, v in bins)
 
+    def fill(self, data, w: NDArray = None, keys: Sequence[str] = None) -> None:
+        """Fill histogram by incrementing bins with data points weighted by w
+
+        Parameters
+        ----------
+        data
+            a ndarray with inner dimension equal to number of axes, or a list
+            of equal-length 1d-arrays containing data for each axis, or a
+            Mapping to 1d-arrays containing data for each axis (requires keys),
+            or a Pandas dataframe (optionally takes a list of keys)
+        w
+            weight to use for incrementing data points. If None, use 1 for all
+        keys
+            list of keys to use if data is a pandas ''DataFrame'' or ''Mapping''
+        """
+        if keys is not None:
+            if isinstance(keys, str):
+                keys = [keys]
+            elif not isinstance(keys, list):
+                keys = list(keys)
+
+        if (
+            isinstance(data, np.ndarray)
+            and len(data.shape) == 1
+            and len(self.binning) == 1
+        ):
+            N = len(data)
+            data = [data]
+        elif (
+            isinstance(data, np.ndarray)
+            and len(data.shape) == 2
+            and data.shape[1] == len(self.binning)
+        ):
+            N = data.shape[0]
+            data = data.T
+        elif isinstance(data, pd.DataFrame) and (
+            (keys is not None and len(keys) == len(self.binning))
```

Comment:
I think what you have is equivalent to this:
```
        elif isinstance(data, pd.DataFrame) and (
            (keys is not None and len(keys) == len(self.binning))
            or data.ndim == len(self.binning)
        ):
```
Unless I'm missing something?

Reply:
I don't think so. My version only checks the dimension of the dataframe itself, when no keys are specified. If keys are specified, only the count of keys is taken into account.

You version always allows for both. Example: in the case of a 3D hsitogram this would allow a user to pass a 3-col dataframe, but only 2 keys. That would obviously not work...

(I failed to attach it to the right position in the code.  My code should replace the line in GitHub _and the following line_). In total my suggestion should read like this (sorry for the confusion).

```python
        [...]
        elif isinstance(data, pd.DataFrame) and (
            len(keys) == len(self.binning)
            if keys is not None
            else data.ndim == len(self.binning)
        ):
            if keys is not None:
                [...]
```

Reply:
From my side, only this point is still open @iguinn. If we address this, the Histogram part should be fine, i suppose!